python/phase-2.md

```diff
@@ -828,4 +829,168 @@ Notes:
+  that requires us to have a two-parameter overload version of `execute_activity` _and_ to make the first param optional
+  like so:
+  
+    ```python
+    @overload
+    async def execute_activity(
+        activity: Callable[[LocalParamType, LocalParamType2], LocalReturnType],
+        arg: Optional[LocalParamType],
+        arg2: LocalParamType2,
+        /,
+        *,
+    ```
+  This means if you had a real two-param activity, you'd now be able to have an optional first parameter. Also, I am not
+  yet confident how this will look when we remove all of these overloads once Mypy gets their `ParamSpec` support.
+* We could have a completely separate call like so:
+
+    ```python
+    workflow.execute_activity_method(ActivityClass.some_activity, "foo")
+    ```
+* Instead of having `@activity.defn` set an attribute on the function (which is then transitively transferred for others
+  using `functools.wraps`), we could have it return something like `ActivityDefinition[ParamTypes, ReturnType]` (which
+  does not get the transitive benefits). The reason we might want this is then we can have `@activity.method_defn` or
+  similar that is smart enough to convert to an activity definition. I have been trying to avoid actually wrapping
+  functions so far.
+* We could support the following with an overload:
+
+    ```python
+    workflow.execute_activity(ActivityClass.some_activity, "foo", method=True)
+    ```
+  This would work with a special overload like:
+
+    ```python
+    @overload
+    async def execute_activity(
+        activity: Callable[[Any, LocalParamType], LocalReturnType],
+        arg: LocalParamType,
+        /,
+        *,
+        method: Literal[True],
+    ```
+  This seems like a fairly sane way, but requires callers pass a special parameter to call
+* Another outstanding question - is there added benefit _at this time_ to support registering a class full of activity
+  methods instead of individually? I am thinking not (yet).
+
+### Workflow Contracts
+
+Sometimes users may want to separate the workflow into a contract and an impl so that they can be separated. We will
+support this approach.
+
+Here is what it looks like with `typing.Protocol`s:
+
+```python
+@workflow.defn
+class MyWorkflow(Protocol):
+    @workflow.run
+    def run(self, arg: str) -> List[str]:
+        ...
+
+@workflow.defn(name=MyWorkflow.__name__)
+class MyWorkflowImpl:
+    @workflow.run
+    def run(self, arg: str) -> List[str]:
+        return [arg]
+
+async def run_workflow(client: Client) -> List[str]:
+    return await client.execute_workflow(MyWorkflow.run)
+
+def create_worker(client: Client) -> Worker:
+    return Worker(workflows=[MyWorkflowImpl])
+```
+
+And here's what it looks like with abstract classes:
+
+```python
+@workflow.defn
+class MyWorkflow(ABC):
+  @workflow.run
+  @abstractmethod
+  def run(self, arg: str) -> List[str]:
+    raise NotImplementedError
+
+@workflow.defn(name=MyWorkflow.__name__)
```

Comment:
Is this required with a subclass?

Reply:
See points below this about annoying inheritance concerns.